File: src/2025-02-24-deepclaude-rust/go/exercises/practice/scale-generator/scale_generator.go
package scale

import "strings"

var sharpNotes = []string{"A", "A#", "B", "C", "C#", "D", "D#", "E", "F", "F#", "G", "G#"}
var flatNotes = []string{"A", "Bb", "B", "C", "Db", "D", "Eb", "E", "F", "Gb", "G", "Ab"}

var sharpTonics = map[string]bool{
	"C":  true,
	"G":  true,
	"D":  true,
	"A":  true,
	"E":  true,
	"B":  true,
	"F#": true,
	"a":  true,
	"e":  true,
	"b":  true,
	"f#": true,
	"c#": true,
	"g#": true,
	"d#": true,
}

var flatTonics = map[string]bool{
	"F":  true,
	"Bb": true,
	"Eb": true,
	"Ab": true,
	"Db": true,
	"Gb": true,
	"d":  true,
	"g":  true,
	"c":  true,
	"f":  true,
	"bb": true,
	"eb": true,
}

func normalizeTonic(tonic string) string {
	if len(tonic) == 0 {
		return ""
	}
	base := string(tonic[0])
	if len(tonic) == 1 {
		return base
	}
	accidental := strings.ToLower(tonic[1:])
	if accidental == "b" || accidental == "#" {
		return base + accidental
	}
	return base + accidental
}

func findIndex(notes []string, tonic string) int {
	for i, note := range notes {
		if note == tonic {
			return i
		}
	}
	return -1
}

func Scale(tonic, interval string) []string {
	normalizedTonic := normalizeTonic(tonic)
	var notes []string
	if sharpTonics[normalizedTonic] {
		notes = sharpNotes
	} else if flatTonics[normalizedTonic] {
		notes = flatNotes
	} else {
		notes = sharpNotes
	}

	index := findIndex(notes, normalizedTonic)
	if index == -1 {
		return nil
	}

	chromatic := append(notes[index:], notes[:index]...)

	if interval == "" {
		return chromatic
	}

	result := []string{chromatic[0]}
	currentIndex := 0
	for _, r := range interval {
		var steps int
		switch r {
		case 'M':
			steps = 2
		case 'm':
			steps = 1
		case 'A':
			steps = 3
		default:
			continue
		}
		currentIndex = (currentIndex + steps) % 12
		result = append(result, chromatic[currentIndex])
	}

	return result
}

File: src/python-2025-02-24-deepclaude-rust/python/exercises/practice/scale-generator/scale_generator.py
class Scale:
    def __init__(self, tonic):
        self.tonic = self.normalize_tonic(tonic)
        self.use_sharps = self.determine_key_signature()
        self.notes = self.sharp_notes if self.use_sharps else self.flat_notes
        self.tonic_index = self.notes.index(self.tonic)
    
    def normalize_tonic(self, tonic):
        if not tonic:
            return tonic
        return tonic[0].upper() + tonic[1:].lower()
    
    def determine_key_signature(self):
        major_sharp = {'G', 'D', 'A', 'E', 'B', 'F#'}
        minor_sharp = {'E', 'B', 'F#', 'C#', 'G#', 'D#'}
        natural_sharp = {'C', 'A'}
        sharp_tonics = major_sharp.union(minor_sharp).union(natural_sharp)
        
        major_flat = {'F', 'Bb', 'Eb', 'Ab', 'Db', 'Gb'}
        minor_flat = {'D', 'G', 'C', 'F', 'Bb', 'Eb'}
        flat_tonics = major_flat.union(minor_flat)
        
        if self.tonic in sharp_tonics:
            return True
        elif self.tonic in flat_tonics:
            return False
        else:
            return True
    
    @property
    def sharp_notes(self):
        return ['C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#', 'A', 'A#', 'B']
    
    @property
    def flat_notes(self):
        return ['C', 'Db', 'D', 'Eb', 'E', 'F', 'Gb', 'G', 'Ab', 'A', 'Bb', 'B']
    
    def chromatic(self):
        return self.notes[self.tonic_index:] + self.notes[:self.tonic_index]
    
    def interval(self, intervals):
        current_index = self.tonic_index
        scale = [self.notes[current_index]]
        step_map = {'m': 1, 'M': 2, 'A': 3}
        for interval in intervals:
            step = step_map.get(interval, 0)
            current_index = (current_index + step) % 12
            scale.append(self.notes[current_index])
        return scale
